Added system.py to physics/mechanics

jbm950 opened the pull request:
Added a class called SymbolicSystem to physics/mechanics to act as a container class for the equations of motion and other details related to dynamic systems. I have created all of the class attributes and docstrings. This work emerged from work done in PR #11182  and in PR [#353](https://github.com/pydy/pydy/pull/353) in the pydy repository.

Would it be better to name the file `symbolicsystem`, `symsystem` or just leave it as `system`?

jbm950 committed:
Added system.py to physics/mechanics

Added a class called SymbolicSystem to physics/mechanics to act as a
container class for the equations of motion and other details related to
dynamic systems. I have created all of the class attributes and
docstrings.

jbm950 committed:
Added system.py to physics/mechanics/__init__.py

Now SymbolicSystem can be called from sympy.physics.mechanics

jbm950 committed:
Added tests for system.p

Tests are now written for system.py and were used when coding up
SymbolicSystem.

jbm950 committed:
Finished coding system.py

A first draft of the code for SymbolicSystem in
sympy/physics/mechanics/system.py is completed.

jbm950 commented:
@moorepants The code and the tests have been added if you'd like to start looking over those.

jbm950 committed:
Added documentation for physics/mechanics/system.py

Documentation has been added for SymbolicSystem. This includes a page
that renders SymbolicSystem's docstrings and a page that goes over an
example of manual entry of equations of motion to SymbolicSystem.

jbm950 commented:
Some documentation has now also been added. The example was adapted from the second example in PR [#353](https://github.com/pydy/pydy/pull/353) in the pydy repository.

jbm950 commented:
@moorepants @asmeurer The error for the tests appears to be in the docstring of the new class I added but I do not know what is wrong. Have either of you seen this error before? It looks like it doesn't like my indentation somewhere?

moorepants commented:
Try running sphinx locally on the docs and see if the error is better. You can also try running rst2html directly on that documentation file to see if any error's pop up.

jbm950 committed:
Added an example to the docstring of SymbolicSystem

I added an example of manually entering the equations of motion to a
SymbolicSystem object to SymbolicSystem's docstring.

angadhn commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+Next step is to define the equations of motion in multiple forms:
+
+    [1] Explicit form where the kinematics and dynamics are combined
+        x' = F(x, t, r, p)
```
On the RHS, I would call this F_1 or some other symbol altogether. I presume that the F here is different from the one immediately below it. The 2 F's also have the same arguments which could lead to some confusion for the reader.

jbm950 replied:
@moorepants I agree with Angadh that this would be a good change. It is worth noting I pulled this section from the code generator in pydy and so the change in notation is equally valid there.

moorepants replied:
That's fine to subscript the M's and F's.

jbm950 committed:
Fixed an error in the docstring of system.py

I needed to pass in the coordinates and speeds as lists instead of sympy
symbols in the example in SymbolicSystem's docstring.

jbm950 committed:
Added clarity to equations in system.py

In the docstring of symsystem there were some equations that repeated
function names where the functions represented a different set of
equations. This as been fixed by adding subscripts to each to signify
that they are different.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+Symbolic Systems in Physics/Mechanics
+=====================================
+
+The `SymbolicSystem` class in physics/mechanics is a location for the pertinent
```
pertinent what?

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+=====================================
+
+The `SymbolicSystem` class in physics/mechanics is a location for the pertinent
+of a dynamic system. In its most basic form it contains the equations of motion
```
should we say "multibody dynamic system" instead of "dynamic system"?

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+The `SymbolicSystem` class in physics/mechanics is a location for the pertinent
+of a dynamic system. In its most basic form it contains the equations of motion
+for the dynamic system, however, it can also contain information regarding the
+loads that the system is feeling, the bodies that the system is comprised of
```
feeling > subject to

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+loads that the system is feeling, the bodies that the system is comprised of
+and any additional equations the user feels is important for the system. The
+goal of this class is to provide a unified output format for the equations of
+motion that simulation code can be designed around.
```
that numerical analysis code can be design around.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+============================
+
+This code will go over the manual input of the equations of motion for the
+simple pendulum into `SymbolicSystem` using x and y coordinates instead of
```
for the simple pendulum that uses the Cartesian location of the mass as the generalized coordinates.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+theta.
+
+The equations of motion are formed at
+http://nbviewer.jupyter.org/github/bmcage/odes/blob/master/docs/ipython/Planar%20Pendulum%20as%20DAE.ipynb ::
```
We also have the eoms  for this in a test for lagranges method. An internal link to that would be a more reliable link.

jbm950 replied:
I found a page in the documentation of sympy mechanics that forms these equations of motion. The example though uses a different orientation of x, y axes than I do and uses q1 and q2 instead of x and y. I put a link to this example (I think, I'm better at working with markdown than rst) and mentioned the formation differences.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+
+where
+
+    x : states, i.e. [q, u]
```
i.e. > e.g.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+        equations or combined dynamics and kinematics)
+    F : right hand side (implicit dynamical equations or combined dynamics
+        and kinematics)
+    G : right hand side of the kinematical differential equations ::
```
in restructured text you can prepend the lines with `|` to ensure each line is on a newline. I'd also move the `::` to its own line.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+Now the bodies and loads iterables need to be initialized. ::
+
+    >>> bodies = [Pa]
+    >>> loads = [(P, g * m * N.x)]
```
Couldn't you do this first in the example and generate the EoMs from LagrangesMethod instead of citing the jupyter notebook?

jbm950 replied:
The idea of the example is to showcase coming up with the equations by hand and entering them manually into `SymbolicSystem`. `LagrangesMethod` should eventually have a method `to_system` and so I don't know that it would be ideal to have the example on `SymbolicSystem`'s page to manually transfer data from `LagrangesMethod`.

moorepants replied:
Is there a wikipedia article with the x, y pendulum derivation that you can point to for the equations of motion? Then you can just say you took the EoMs from there.

moorepants replied:
Also, it doesn't really make much sense to have the bodies and loads manually created if you haven't set all the positions and velocities manually.

jbm950 replied:
The wikipedia article for [pendulum math](https://en.wikipedia.org/wiki/Pendulum_(mathematics)) does not seem to have the derivation for the x, y, lambda version of the system and doing a google search for "pendulum x and y coordinates" did not produce promising results either. I went ahead and set the velocities

jbm950 committed:
Fixed documentation errors

I believe this will fix the indentation errors that were being produced
from the documentation of system.py

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+class. ::
+
+    >>> alg_con = [2]
+    >>> alg_con_full = [4]
```
Do you have any idea if we could just check the equations internally? Is it easy to check if a given expression is algebraic or not? If so, it'd be nice to avoid this, especially since it is a little confusing wrt to input/output differences.

jbm950 replied:
I think you could determine whether or not there are algebraic equations present by checking the number of kinematical differential equations and the number of coordinates. Overall though I do not see a method of determining which of the coordinates represent derivatives and which ones do not.

moorepants replied:
Would it simply be that there are zeros in the mass matrix for the associated equations?

jbm950 replied:
No because M*u where u includes lambda so there would be an algebraic equation in the row multiplied by lambda. Lamba has it's own rows in LagrangesMethod (and is merged with the mass matrix) so while forming the `SymbolicSystem` there the information could be pulled and passed in to `SymbolicSystem` as above.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+    >>> alg_con = [2]
+    >>> alg_con_full = [4]
+
+An iterable containing the states now needs to be created for the solvers. ::
```
What "solvers" is this referring to?

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+    ...                                    mass_matrix=dyn_implicit_mat,
+    ...                                    coordinate_derivatives=kin_explicit_rhs,
+    ...                                    alg_con=alg_con, coord_idxs=(0, 1), 
+    ...                                    speed_idxs=(2, 3))
```
It'd be better to say what `coord_idxs` and `speed_idxs` are before using them.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,184 @@
+    >>> symsystem1.dynamic_symbols
+    set([lambda(t), u(t), v(t), x(t), y(t)])
+    >>> symsystem1.constant_symbols
+    set([g, l, m])
```
I think it was a bad choice for us to use sets in pydy.system.System for these. A tuple is nicer to work with because you can index it. A list would be fine too.

jbm950 replied:
The set is being used to eliminate duplicates. I could change it with an added if statement to determine if each item is already in the list. Is this the preferred course of action?

moorepants replied:
I don't care what is done internally to the class, but the user accessible attribute should return a tuple or list, not a set.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,437 @@
+    or combined dynamics and kinematics)
+    F : right hand side (implicit dynamical equations or combined dynamics and
+    kinematics)
+    G : right hand side of the kinematical differential equations
```
This list may render incorrectly with Sphinx.

jbm950 replied:
I'll change the indentation and see if the errors reappear. It might have just been in the .rst file it didn't like the indentation and this was the only way I could get rid of the error.

moorepants replied:
In general you can't just write new lines like this an expect them to render correctly in rst, you have to do something like:

```
| x : stuff
| y: stuff
| z: stuff
```

jbm950 replied:
This is fixed by naming the F's and M's explicitly as is now also done in the example.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,437 @@
+            x' = F_1(x, t, r, p)
+
+        [2] Implicit form where the kinematics and dynamics are combined
+            M(x, p) x' = F_2(x, t, r, p)
```
If you are doing F_1, F_2, F_3, you should have M_2 and M_3 also.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,437 @@
+    ==========
+
+    TODO - come up with the size of the matrices and add a notes section
+    detailing any variables I used
```
Needs to be done.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,437 @@
+        (reference frame acted upon, torque vector).
+
+    Parameters
+    ==========
```
I prefer the parameters section to be in the docstring for `__init__`.

jbm950 replied:
I agree but it won't be picked up by autodoc that way.

moorepants replied:
Ah, there is a setting that you can list in sphinx to get the autodoc to grab this:

http://stackoverflow.com/questions/5599254/how-to-use-sphinxs-autodoc-to-document-a-classs-init-self-method

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,437 @@
+    def coordinates(self):
+        """Returns the column matrix of the generalized coordinates"""
+        if self._coordinates is None:
+            raise AttributeError("The coordinates were not specified")
```
periods at the end of these error sentences.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,437 @@
+    @property
+    def states(self):
+        """Returns the column matrix of the state variables"""
+        return self._states
```
There is no reason for these to be properties if they don't have anything special going on in the setters or getters.

jbm950 replied:
These have no setter and thus by making them properties errors will arise if the user accidentally tries to overwrite them. Just like the properties `dependent_speeds` and `independent_speeds` in the [KanesMethod rework](https://github.com/sympy/sympy/pull/11183/files).

moorepants replied:
Ok, these are not writable. Thanks.

jbm950 committed:
Revised documentation based on reviewer comments

I went through and made changes in the docstring and example
documentation for system to address reviewer comments.

moorepants commented on doc/src/modules/physics/mechanics/symsystem.rst:
```diff
@@ -0,0 +1,196 @@
+    [3] Implicit form where the kinematics and dynamics are separate
+        M_3(q, p) u' = F_3(q, u, t, r, p)
+        q' = G(q, u, t, r, p)
+
```
I think it would be nice to introduce the variable naming convention we have in the class near these definitions so that it is easily found and understood. Maybe in the list below you can note what attribute name is associated with what variable in these 3 equations.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+
+
+    Attributes
+    ==========
```
It would be nice to have the variable name from your equation forms in these descriptions so that people can connect the attribute name to the equations.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+
+    dynamic_symbols : Matrix
+        This matrix contains all of the symbols that have a time dependence in
+        the equations of motion of the system
```
Mention that these will have both the state variables, lagrange multipliers, and exogenous inputs.

moorepants replied:
They are also not technically symbols, but sympy functions of time.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+        If the loads in the system were given they are stored in a list for
+        future access. This includes forces and torques where forces are given
+        by (point of application, force vector) and torques are given by
+        (reference frame acted upon, torque vector).
```
Is a tuple preferable for these? Any reason that the user should be able to mutate the lists? If not, maybe tuple is better.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+                inter1 = eye(num_kin_eqns).row_join(zeros1)
+                inter2 = zeros2.row_join(self._dyn_implicit_mat)
+                self._comb_implicit_mat = inter1.col_join(inter2)
+                return self._comb_implicit_mat
```
Can all this go in the initializer instead of here? It'd be nice, in general, if the properties don't really do much of any computation.

moorepants replied:
Well, this is probably ok. It is light enough.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+        is to provided upon initialization, this method will calculate it. This
+        calculation can potentially take awhile to compute."""
+        inter = self._comb_implicit_mat.LUsolve(self._comb_implicit_rhs)
+        self._comb_explicit_rhs = inter
```
In general, you only ever have to invert the mass matrix. This seems to be inverting the "full mass matrix". I'm pretty sure the `KanesMethod.rhs()` function does this efficiently now:

https://github.com/sympy/sympy/blob/master/sympy/physics/mechanics/kane.py#L743

jbm950 replied:
In general you may not know which form the equations of motion are given in and so I believe that avoiding the kinematical equations is only possible some of the time. (See below on the kinematical equations). I will write a conditional to use the more efficient calculation when possible.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+        form, x' = F, where the kinematical equations are included"""
+        if self._comb_explicit_rhs is None:
+            raise AttributeError("Please run .combute_explicit_form before "
+                                 "attempting to access comb_explicit_rhs.")
```
This seems like a reasonable solution.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+            raise AttributeError("kin_explicit_rhs is not specified for "
+                                 "equations of motion form [1] or [2].")
+        else:
+            return self._kin_explicit_rhs
```
Shouldn't this extract the equations from the full mass matrix and full forcing form?

jbm950 replied:
How would you suggest extracting it from the combined implicit form of the equations? Keep in mind in general the user only has to specify the states. If you do know coordinates and speeds information could you pull off the first number of speeds rows from the right hand side and the equivalent sized matrix from the mass matrix and then invert and multiply them? Would this work in a general? Would the inverting potentially be costly?

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+                find_dynamicsymbols(expr))
+        functions_of_time = functions_of_time.union(self._states)
+
+        return functions_of_time
```
I recommend `return tuple(functions_of_time)`.

I think this should be a method, no? This computation can take some time for big expressions.

jbm950 replied:
I have changed it to `return tuple(functions_of_time)` but will this be useful given that the order is randomized every time? I'm going to have to remove showing the method from symsystem.rst because the output is in a different order every time and so the doctest will never pass. I went ahead and changed it to a method as well.

The way it returns now with the change the user will be able to index the results but will not know the order of the results ahead of time. An alternative would be 

```python
if item not in list:
    list.append(item)
```

I think this would result in consistent output, what do you think?

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+        """Returns a dictionary of the given output equations where the key's
+        are user defined names for each equation and the values are the symbolic
+        equations themselves"""
+        return self._output_eqns
```
Is a SymbolicSystem intended to be immutable? It seems reasonable that this could be a writable attribute.

moorepants commented on sympy/physics/mechanics/system.py:
```diff
@@ -0,0 +1,441 @@
+        if self._loads is None:
+            raise AttributeError("loads were not specified for the system.")
+        else:
+            return self._loads
```
I recommend returning a tuple for bodies and loads.

moorepants commented on sympy/physics/mechanics/tests/test_system.py:
```diff
@@ -0,0 +1,238 @@
+    # Attribute error when trying to access comb_explicit_rhs before it was
+    # calculated
+    with raises(AttributeError):
+        symsystem2.comb_explicit_rhs
```
Can you see how well this coverage is?

https://github.com/sympy/sympy/blob/master/bin/coverage_report.py

moorepants commented:
I've reviewed this, left a number of minor comments. This is looking really good. Very close to merge-able.

moorepants commented:
@angadhn How does this one look to you? We'd like to get it in soon.

And, do you have any thoughts on being able to automatically identify any algebraic equations, so we could potentially eliminate the need for the user to specify which equations are algebraic.

jbm950 committed:
Made some major revisions to address reviewer comments

There are many changes throughout the system.py code in order to address
reviewer comments.

jbm950 commented:
<img width="763" alt="screen shot 2016-08-10 at 12 22 16 pm" src="https://cloud.githubusercontent.com/assets/12311558/17561656/1c248e18-5ef5-11e6-8d32-d8b153d1c56c.png">

jbm950 committed:
I think this may fix the sphinx error

There is an error in sphinx pulling the __init__ method from
SymbolicSystem. This commit hopefully fixes that.

jbm950 commented:
Attempting to autodoc the __init__ method is causing travis errors, may I move the parameters section back to the general docstring?

moorepants commented:
That's fine.

moorepants commented:
Nice coverage on system. Can't get better than that.

jbm950 committed:
Moved Parameters documentation back to main docstring

There was an attempt at autodocumenting the __init__ method of system
but it wasnt working and so the parameters documentation was moved back
to the main docstring.

moorepants commented:
+1, merging this.
